src/axom/quest/Shaper.cpp

```diff
@@ -30,13 +27,133 @@ constexpr double Shaper::MINIMUM_PERCENT_ERROR;
 constexpr double Shaper::MAXIMUM_PERCENT_ERROR;
 constexpr double Shaper::DEFAULT_VERTEX_WELD_THRESHOLD;
 
-Shaper::Shaper(const klee::ShapeSet& shapeSet, sidre::MFEMSidreDataCollection* dc)
-  : m_shapeSet(shapeSet)
+#if defined(AXOM_USE_MFEM)
+Shaper::Shaper(RuntimePolicy execPolicy,
+               int allocatorId,
+               const klee::ShapeSet& shapeSet,
+               sidre::MFEMSidreDataCollection* dc)
+  : m_execPolicy(execPolicy)
+  , m_allocatorId(allocatorId != axom::INVALID_ALLOCATOR_ID
+                    ? allocatorId
+                    : axom::policyToDefaultAllocatorID(execPolicy))
+  , m_shapeSet(shapeSet)
   , m_dc(dc)
+  #if defined(AXOM_USE_CONDUIT)
+  , m_bpGrp(nullptr)
+  , m_bpTopo()
+  , m_bpNodeExt(nullptr)
+  , m_bpNodeInt()
+  #endif
 {
-#if defined(AXOM_USE_MPI) && defined(MFEM_USE_MPI)
+  #if defined(AXOM_USE_MPI) && defined(MFEM_USE_MPI)
   m_comm = m_dc->GetComm();
+  #endif
+  m_cellCount = m_dc->GetMesh()->GetNE();
+
+  setFilePath(shapeSet.getPath());
+}
+#endif
+
+Shaper::Shaper(RuntimePolicy execPolicy,
+               int allocatorId,
+               const klee::ShapeSet& shapeSet,
+               sidre::Group* bpGrp,
+               const std::string& topo)
+  : m_execPolicy(execPolicy)
+  , m_allocatorId(allocatorId != axom::INVALID_ALLOCATOR_ID
+                    ? allocatorId
+                    : axom::policyToDefaultAllocatorID(execPolicy))
+  , m_shapeSet(shapeSet)
+#if defined(AXOM_USE_CONDUIT)
+  , m_bpGrp(bpGrp)
+  , m_bpTopo(topo.empty() ? bpGrp->getGroup("topologies")->getGroupName(0) : topo)
+  , m_bpNodeExt(nullptr)
+  , m_bpNodeInt()
+#endif
+  , m_comm(MPI_COMM_WORLD)
+{
+  SLIC_ASSERT(m_bpTopo != sidre::InvalidName);
+
+  // This may take too long if there are repeated construction.
+  m_bpGrp->createNativeLayout(m_bpNodeInt);
+
+#if defined(AXOM_DEBUG)
+  std::string whyBad;
+  bool goodMesh = verifyInputMesh(whyBad);
+  SLIC_ASSERT_MSG(goodMesh, whyBad);
+#endif
+
+  m_cellCount = conduit::blueprint::mesh::topology::length(
+    m_bpNodeInt.fetch_existing("topologies").fetch_existing(m_bpTopo));
+
+  setFilePath(shapeSet.getPath());
+}
+
+Shaper::Shaper(RuntimePolicy execPolicy,
+               int allocatorId,
+               const klee::ShapeSet& shapeSet,
+               conduit::Node* bpNode,
```

Comment:
bpNode could be passed a nullptr and get used before getting to the method body where we could assert.
Also, since core contains a method to get the allocatorID from the RuntimePolicy now, is there a reason to have to pass allocatorId here?

Reply:
Will change bpNode to a reference.  I actually prefer references where we don't want to imply that a nullptr is meaningful, but I was following existing examples.  The allocator id is for flexibility, as @kennyweiss stated.